scalding-core/src/main/scala/com/twitter/scalding/Execution.scala

```diff
@@ -380,70 +387,78 @@ object Execution {
     def cleanCache: EvalCache =
       new EvalCache(writer)
 
-    def getOrLock(cfg: Config, write: ToWrite[_]): Either[Promise[Counters], Future[Counters]] =
+    def getOrLock(cfg: Config, write: ToWrite[_]): (Either[Promise[Counters], Future[Counters]], CancellationHandler) =
       toWriteCache.getOrPromise((cfg, write))
 
     def getOrElseInsertWithFeedback[T](cfg: Config, ex: Execution[T],
-      res: => Future[(T, Counters)]): (Boolean, Future[(T, Counters)]) =
+      res: => (Future[(T, Counters)], CancellationHandler)): (Boolean, (Future[(T, Counters)], CancellationHandler)) =
       // This cast is safe because we always insert with match T types
       cache.getOrElseUpdateIsNew((cfg, ex), res)
-        .asInstanceOf[(Boolean, Future[(T, Counters)])]
+        .asInstanceOf[(Boolean, (Future[(T, Counters)], CancellationHandler))]
 
     def getOrElseInsert[T](cfg: Config, ex: Execution[T],
-      res: => Future[(T, Counters)]): Future[(T, Counters)] =
+      res: => (Future[(T, Counters)], CancellationHandler)): (Future[(T, Counters)], CancellationHandler) =
       getOrElseInsertWithFeedback(cfg, ex, res)._2
   }
 
 
   private[scalding] final case class FutureConst[T](get: ConcurrentExecutionContext => Future[T]) extends Execution[T] {
     protected def runStats(conf: Config, mode: Mode, cache: EvalCache)(implicit cec: ConcurrentExecutionContext) =
-      Trampoline(cache.getOrElseInsert(conf, this,
-        for {
+      Trampoline {
+        lazy val fut = for {
           futt <- toFuture(Try(get(cec)))
           t <- futt
-        } yield (t, Map.empty)))
+        } yield (t, Map.empty[Long, ExecutionCounters])
+
+        cache.getOrElseInsert(conf, this, (fut, CancellationHandler.empty))
```

Comment:
could we not use:

https://github.com/twitter/scalding/pull/1920/files#diff-122dc0ec8aa7eec7186fd18a43ec7d67R24

here?